// shopee-ui/src/base/directives/clickoutside.js
const ctx = 'shopee-ui-clickoutside';

let nodeId = 0;
let nodes = [];
let listening = false;

const handleClick = e => {
  nodes.forEach(node => {
    if (!node.contains(e.target)) {
      if (node[ctx].expressionFn) {
        node[ctx].expressionFn(e);
      } else {
        node[ctx].bindingFn && node[ctx].bindingFn(e);
      }
    }
  });
};

/**
 * trigger callback when click outside the binding element
 */
export default {
  bind(el, binding, vnode) {
    el[ctx] = {
      id: nodeId++,
      bindingFn: binding.value,
      expressionFn: vnode.context[binding.expression]
    };
    nodes.push(el);
    if (!listening) {
      document.addEventListener('click', handleClick);
      listening = true;
    }
  },
  update(el, binding, vnode) {
    el[ctx].expressionFn = vnode.context[binding.expression];
    el[ctx].bindingFn = binding.value;
  },
  unbind(el) {
    for (let i = 0, j = nodes.length; i < j; i++) {
      if (nodes[i][ctx].id === el[ctx].id) {
        nodes.splice(i, 1);
        break;
      }
    }
    if (nodes.length === 0) {
      document.removeEventListener('click', handleClick);
      listening = false;
    }
  }
};
