// shopee-ui/packages/row/src/row.js
export default {
  name: 'ShopeeRow',

  props: {
    tag: {
      type: String,
      default: 'div'
    },
    gutter: {
      type: Number,
      default: 16
    },
    type: String,
    justify: {
      type: String,
      default: 'start',
      validator(justify = 'start') {
        return ['start','end','center','space-around','space-between'].indexOf(justify) >= 0;
      }
    },
    align: {
      type: String,
      default: 'top',
      validator(align = 'top') {
        return ['top','middle','bottom'].indexOf(align) >= 0;
      }
    }
  },

  computed: {
    style() {
      const ret = {};

      if (this.gutter) {
        ret.marginLeft = `-${this.gutter / 2}px`;
        ret.marginRight = ret.marginLeft;
      }

      return ret;
    }
  },

  render(h) {
    return h(this.tag, {
      class: [
        'shopee-row',
        this.justify !== 'start' ? `is-justify-${this.justify}` : '',
        this.align !== 'top' ? `is-align-${this.align}` : '',
        this.type === 'flex' ? 'shopee-row__flex' : ''
      ],
      style: this.style
    }, this.$slots.default);
  }
};
